### DISCUSSION TITLE
No usable Android build tools found. Highest 33.x installed version is 33.0.0; Recommended version is 33.0.2

### DISCUSSION THREAD

**Discussion by BorisB** [codemagic-ci-cd/codemagic-docs#1989] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Ionic/Cordova

### Steps to reproduce

1. Create cordova app
2. Add codemagic.yml with workflow
```
  android-workflow:
    name: Android workflow
    environment:
      groups:
        # Add the group environment variables in Codemagic UI (either in Application/Team variables) - https://docs.codemagic.io/variables/environment-variable-groups/
        - keystore_credentials # <-- (Includes KEYSTORE, KEYSTORE_PASSWORD, KEY_ALIAS_PASSWORD, KEY_ALIAS)
          # KEYSTORE <-- Put your keystore file here
          # KEYSTORE_PASSWORD <-- Put your keystore password here
        # KEY_ALIAS_PASSWORD <-- Put your keystore alias password here
        # KEY_ALIAS <-- Put your keystore alias username here
        - google_play # <-- (Includes GCLOUD_SERVICE_ACCOUNT_CREDENTIALS - Put your google-services.json here)
        - app
      vars:
        KEYSTORE_PATH: '/tmp/keystore.keystore'
      xcode: latest
      node: v16.16.0
      npm: 8
    cache:
      cache_paths:
        - $CM_BUILD_DIR/node_modules
    scripts:
      - name: Install dependencies
        script: |
          npm install --location=global cordova@11.0.0
          npm ci
          npm install --location=global cordova-set-version
          npx cordova-set-version -v $VERSION -b $BUILD_NUMBER
      - name: Add Android platform
        script: |
          set -x
          cordova platform remove android --nosave
          cordova platform add android --confirm --no-interactive --noresources
      - name: Build Android
        script: |
          set -x
          set -e
          cordova build android --release --no-interactive --prod --device
          echo $KEYSTORE | base64 --decode > $KEYSTORE_PATH
          UNSIGNED_APK_PATH=$(find platforms/android/app/build/outputs -name "*.aab" | head -1)
          jarsigner -keystore "${KEYSTORE_PATH}" -storepass "${KEYSTORE_PASSWORD}" -keypass "${KEY_ALIAS_PASSWORD}" "${UNSIGNED_APK_PATH}" "${KEY_ALIAS}"
          mv $UNSIGNED_APK_PATH $(echo $UNSIGNED_APK_PATH | sed 's/-unsigned//')
    artifacts:
      - platforms/android/app/build/outputs/**/*.aab
      - platforms/android/app/build/outputs/**/mapping.txt
```
3. Start new build

### Expected results

Build successed

### Actual results

FAILURE: Build failed with an exception.

* Where:
Script '/Users/builder/clone/platforms/********/CordovaLib/cordova.gradle' line: 75

* What went wrong:
A problem occurred evaluating script.
> No usable Android build tools found. Highest 33.x installed version is 33.0.0; Recommended version is 33.0.2.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.6/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 4s
Command failed with exit code 1: /Users/builder/clone/platforms/********/gradlew -b /Users/builder/clone/platforms/********/build.gradle :app:bundleRelease

Build failed :|
Step 4 script `Build Android` exited with status code 1

### Build id (optional)

64e3d8822eb0f2b7c1839708

**Comment by kulakarla** [comment]
Hello @BorisB 

Can you [VNC connect](https://docs.codemagic.io/troubleshooting/accessing-builder-machine-via-ssh/) to the build machine, open Android Studio, install the necessary Android Build Tools version (33.0.2) and try running the build command again.

If the build is successful then we'll proceed with necessary changes to our build machines.

**Reply by BorisB** [reply]
I confirm that after `/usr/local/share/android-sdk/cmdline-tools/latest/bin/sdkmanager "build-tools;33.0.2"` build was successful.

**Reply by kulakarla** [reply, marked as answer]
Run this script before you build to install the right build tools automatically every time. Due to it being a fairly niche Ionic issue, we'll not be updating all machines just yet.
```
$ANDROID_SDK_ROOT/tools/bin/sdkmanager --install "build-tools;33.0.2
```

If you encounter a Java XmlSchema error, use an older version. E.g in the YAML under environment, declare `java: 1.8` to use Java 8.

**Reply by BorisB** [reply]
I would suggest to add this to example in documentation.

**Reply by kulakarla** [reply]
Thank you for your suggestion, noted. I will now be closing this thread.

**Reply by L3VU5** [reply]
I had to use the script above and set the `java: 1.8` not to have a Java XmlSchema error. It successfully finished the script, but at the later stage, the `react-native-firebase_messaging` is failing during Android build release, because it's not compatible with the java 1.8.

```
> Task :react-native-firebase_messaging:compileReleaseJavaWithJavac FAILED
An exception has occurred in the compiler (1.8.0_382).
```

**Comment by brenosprlgc** [comment]
@kulakarla is that possible change gradle version ? Because cordova 12 uses gradle 7.6.
